File: api-chi/cmd/routes/movie_director_test.go
package routes

import (
	"api-chi/cmd/models"
	"api-chi/internal/message"

	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/go-chi/chi/v5"
	"github.com/stretchr/testify/assert"
)

func Test_MovieDirectorRoutes(t *testing.T) {
	r := chi.NewRouter()
	MovieDirectorRoutes(r)
	id := ""

	t.Run("Count success", func(t *testing.T) {
		req := httptest.NewRequest("GET", "/movie-director/count", nil)
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		var response message.Response
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.GET_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)
	})

	t.Run("GetAll success", func(t *testing.T) {
		search := ""
		limit := 10
		page := 1

		req := httptest.NewRequest("GET", fmt.Sprintf("/movie-director?search=%s&limit=%d&page=%d", search, limit, page), nil)
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		var response message.Response
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.GET_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)
	})

	t.Run("Create success", func(t *testing.T) {
		input := models.MovieDirector{Name: "test director"}
		body, _ := json.Marshal(input)

		req := httptest.NewRequest("POST", "/movie-director", bytes.NewBuffer(body))
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		var response message.Response
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.CREATE_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)

		idValue, ok := response.Data.(string)
		if !ok {
			t.Fatalf("Expected response.Data to be of type string, got %T", response.Data)
		}
		id = idValue
	})

	t.Run("Update success", func(t *testing.T) {
		if id == "" {
			t.Fatal("ID must be set before running Update test")
		}

		input := models.MovieDirector{
			Id:   id,
			Name: "updated director",
		}
		body, _ := json.Marshal(input)

		req := httptest.NewRequest("PATCH", "/movie-director", bytes.NewBuffer(body))
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		var response message.Response
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.UPDATE_DATA_SUCCESS, response.Message)
		assert.NotNil(t, response.Data)
	})

	t.Run("Remove success", func(t *testing.T) {
		if id == "" {
			t.Fatal("ID must be set before running Remove test")
		}

		req := httptest.NewRequest("DELETE", "/movie-director/"+id, nil)
		res := httptest.NewRecorder()

		r.ServeHTTP(res, req)

		assert.Equal(t, http.StatusOK, res.Code)
		var response message.Response
		err := json.NewDecoder(res.Body).Decode(&response)
		assert.NoError(t, err)
		assert.Equal(t, message.REMOVE_DATA_SUCCESS, response.Message)
	})
}

File: api-fastify/src/tests/routes/movie_director.test.ts
import fastify from "fastify"
import MovieDirectorRoutes from "@/api/routes/movie_director"
import { MovieDirector } from "@/types/movie_director"
import { HttpStatus } from "@/types/http_status"

describe("MovieDirectorRoutes", () => {
  let app: ReturnType<typeof fastify>
  let id: string = ""

  beforeAll(async () => {
    app = fastify()
    app.register(MovieDirectorRoutes)
    await app.ready()
  })

  afterAll(async () => {
    await app.close()
  })

  describe("POST /movie-directors", () => {
    it("should create a movie director", async () => {
      const input: MovieDirector = {
        id: "",
        name: "Test Created Movie Director",
      }

      const response = await app.inject({
        method: "POST",
        url: "/movie-directors",
        payload: input,
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      expect(typeof result.data.id).toBe("string")
      expect(typeof result.data.name).toBe("string")
      expect(result.data.name).toBe(input.name)
      id = result.data.id
    })
  })

  describe("PATCH /movie-directors", () => {
    it("should update a movie director", async () => {
      const input: MovieDirector = { id: id, name: "Test Updated Movie Director" }

      const response = await app.inject({
        method: "PATCH",
        url: "/movie-directors",
        payload: input,
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      expect(typeof result.data.id).toBe("string")
      expect(typeof result.data.name).toBe("string")
      expect(result.data.id).toBe(input.id)
      expect(result.data.name).toBe(input.name)
    })
  })

  describe("DELETE /movie-directors/:id", () => {
    it("should remove a movie director", async () => {
      const response = await app.inject({
        method: "DELETE",
        url: `/movie-directors/${id}`,
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
    })
  })

  describe("GET /movie-directors/count", () => {
    it("should count movie directors", async () => {
      const response = await app.inject({
        method: "GET",
        url: "/movie-directors/count",
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      expect(typeof result.data).toBe("number")
    })
  })

  describe("GET /movie-directors", () => {
    it("should get all movie directors", async () => {
      const response = await app.inject({
        method: "GET",
        url: "/movie-directors?search=&limit=10&page=1",
      })

      expect(response.statusCode).toBe(HttpStatus.OK)
      const result = response.json()
      expect(result.success).toBe(true)
      let countData: number = 0
      result.data.forEach((movie_director: MovieDirector) => {
        expect(typeof movie_director.id).toBe("string")
        expect(typeof movie_director.name).toBe("string")
        expect(movie_director.id).not.toBe("")
        expect(movie_director.name).not.toBe("")
        countData += 1
      })
      expect(countData).toBe(10)
    })
  })
})
